### ISSUE TITLE
Downloading workshop mods from within RimSort when there is no SteamDB supplied does not currently work

### ISSUE THREAD

**Issue by jefferyharrell** [RimSort/RimSort#188]
As of commit `effd97c` the functionality to download Workshop mods from within RimSort does not work on macOS 14 or Windows 11. This also applies to `Edge bafb528`.

Steps to reproduce:

1. Delete the RimSort folder. On the Mac, `~/Library/Application Support/RimSort`. On Windows it's under `%APPDATA%/Local/RimSort` I think. This gets us to a fresh environment.
2. Run RimSort from commit `effd97c`.
3. Go through the usual out-of-box experience.
4. Set up `steamcmd`.
5. Click "Browse Workshop".
6. Pick a mod. I used "Corpse Cleanup" for all my testing.
7. Click "Add to List".
8. Click "Download mod(s) (SteamCMD)".
9. Nothing happens in the UI.

The following exception gets written to `stdout` or `stderr`:

```
[2023-10-13 08:03:53] ERROR: Uncaught exception:
Traceback (most recent call last):
  File "/Users/jefferyharrell/Documents/RimWorld_Modding/GitHub/RimSort/view/main_content_panel.py", line 1893, in _do_download_mods_with_steamcmd
    publishedfileids = check_if_pfids_blacklisted(
        publishedfileids=publishedfileids,
        steamdb=self.metadata_manager.external_steam_metadata,
    )
     |-> publishedfileids = ['3047191756']
     |-> check_if_pfids_blacklisted = <function check_if_pfids_blacklisted at 0x16d7b2550>
     |-> self.metadata_manager.external_steam_metadata = None
     => self = <view.main_content_panel.MainContent(0x6000020f1fa0) at 0x16d894c80>

  File "/Users/jefferyharrell/Documents/RimWorld_Modding/GitHub/RimSort/util/metadata.py", line 2113, in check_if_pfids_blacklisted
    if steamdb.get(publishedfileid, {}).get("blacklist"):
     |-> steamdb.get = '!!! Not Exists'
     |-> publishedfileid = '3047191756'
     => publishedfileids = ['3047191756']
     => steamdb = None
     => blacklisted_mods = {}
AttributeError: 'NoneType' object has no attribute 'get'
```

[RimSort.log](https://github.com/RimSort/RimSort/files/12895020/RimSort.log)

**Comment by Ant-V** [comment]
Additional info discovered: 
• It appears to be an issue with steamdb not knowing what to do when the external DB is set to none. 
• This is not an issue in v.1.0.6.2-hf

**Comment by LionelColaso** [comment]
yeah same happens on windows if SteamDB is set to None
![image](https://github.com/RimSort/RimSort/assets/2488703/367d3eda-8edd-4b54-b8b5-d7a8dc43edd4)

**Comment by jefferyharrell** [comment]
Can someone explain to me in simple words that an old man can understand what "Steam Workshop DB" is and what it's supposed to do? I've had a hard time figuring it out from the app's source code.

**Comment by LionelColaso** [comment]
Steamworks is used to interact directly with a running Steam client. This works cross-platform to interface RimSort with Steam on the local computer by issuing native Steamworks API calls to the Steam client.

RimSort uses my personal fork of a Python module called SteamworksPy to include changes to support a few different things which you can find here: https://github.com/twstagg/SteamworksPy

This custom fork adds GetAppDependencies support, minor fixes, and Nuitka support

RimSort uses SteamworksPy -> Steamworks API  calls for the following things:

1) AppDependenciesQuery used in DB Builder to get AppID dependency metadata to add DLC dependencies to mods in the SteamDB 

https://partner.steamgames.com/doc/api/ISteamUGC#GetAppDependencies

https://github.com/RimSort/RimSort/blob/main/util/steam/steamworks/wrapper.py#L152

2. Subscription actions used by RimSort for subscribe/unscubscribe to/from mods via Steam client

https://partner.steamgames.com/doc/api/ISteamUGC#SubscribeItem

https://partner.steamgames.com/doc/api/ISteamUGC#UnsubscribeItem

https://github.com/RimSort/RimSort/blob/main/util/steam/steamworks/wrapper.py#L232

3. RimSort will also launch the game using Steamworks, allowing the game to launch properly with Steam API, Steam overlay, cloud saves,shift + tab, etc

https://github.com/RimSort/RimSort/blob/main/util/steam/steamworks/wrapper.py#L204

**Comment by jefferyharrell** [comment]
Okay @LionelColaso, understood, and thanks for that. But I still don't understand what "Steam Workshop DB" is supposed to be. As far as I can tell RimSort runs just fine without it, so I'm not sure what its purpose is. Do you see what I mean?

**Comment by LionelColaso** [comment]
i guess you got the link on discord just putting it here incase someone gets curious 
https://github.com/RimSort/RimSort/wiki/User-Guide#external-metadata

**Comment by jefferyharrell** [comment]
That section of the wiki is well and good, but it's a little on the dense side and not that easy to parse. What I got out of it is that "Steam Workshop DB" consists of these things:

<img width="305" alt="image" src="https://github.com/RimSort/RimSort/assets/8659114/306ff0b2-df78-44c6-a015-01bda65e97d6">

Do we have any idea why the reported malfunction happens without a Steam Workshop DB but _doesn't_ happen _with_ one?

I've made a blind-idiot attempt at a fix; it's available on my branch here: https://github.com/jefferyharrell/RimSort/tree/jefferyharrell/download-with-steamcmd-fix Would somebody mind looking at it and telling me if it's a terrible idea? It seems to "fix" the problem on macOS and Windows in the sense that the exception doesn't get thrown any more.

**Comment by jefferyharrell** [comment]
Oh, I suppose I should share the patch.

[20aa9f.patch](https://github.com/RimSort/RimSort/files/12895789/20aa9f.patch)

**Comment by LionelColaso** [comment]
I tested this and it works https://github.com/jefferyharrell/RimSort/commit/20aa9f7ea2e9241866021708627fa078f42e57b6

**Comment by twstagg** [comment]
@jefferyharrell I am OK with this patch - feel free to add to metadata-manager

**Comment by jefferyharrell** [comment]
Fixed by #191
